java/ql/lib/semmle/code/java/controlflow/Guards.qll

```diff
@@ -72,6 +73,51 @@ class ConditionBlock extends BasicBlock {
   }
 }
 
+// Join order engineering -- first determine the switch block and the case indices required, then retrieve them.
+bindingset[switch, i]
+pragma[inline_late]
+private predicate isNthCaseOf(SwitchBlock switch, SwitchCase c, int i) { c.isNthCaseOf(switch, i) }
+
+/**
+ * Gets a switch case >= pred, up to but not including `pred`'s successor pattern case,
+ * where `pred` is declared on `switch`.
+ */
+private SwitchCase getACaseUpToNextPattern(PatternCase pred, SwitchBlock switch) {
+  // Note we do include `case null, default` (as well as plain old `default`) here.
+  not result.(ConstCase).getValue(_) instanceof NullLiteral and
+  exists(int maxCaseIndex |
+    switch = pred.getParent() and
+    if exists(getNextPatternCase(pred))
+    then maxCaseIndex = getNextPatternCase(pred).getCaseIndex() - 1
+    else maxCaseIndex = lastCaseIndex(switch)
+  |
+    isNthCaseOf(switch, result, [pred.getCaseIndex() .. maxCaseIndex])
+  )
+}
+
+/**
+ * Gets the closest pattern case preceding `case`, including `case` itself, if any.
+ */
+private PatternCase getClosestPrecedingPatternCase(SwitchCase case) {
+  case = getACaseUpToNextPattern(result, _)
+}
+
+/**
+ * Holds if `pred` is a control-flow predecessor of switch case `sc` that is not a
+ * fall-through from a previous case.
+ *
+ * For classic switches that means flow from the selector expression; for switches
+ * involving pattern cases it can also mean flow from a previous pattern case's type
+ * test or guard failing and proceeding to then consider subsequent cases.
+ */
+private predicate isNonFallThroughPredecessor(SwitchCase sc, ControlFlowNode pred) {
```

Comment:
Based on the qldoc, we should add `pred = sc.getControlFlowNode().getAPredecessor()`.